### ISSUE TITLE
Instantiating a template with new-app and template file doesn't work properly

### ISSUE THREAD

**Issue by siamaksade** [openshift/origin#9314]
I have a template file in yaml:
https://github.com/OpenShiftDemos/openshift-cd-demo/blob/master/cicd-gogs-template.yaml

If I use `oc new-app /path/template.yaml` to instantiate the template, resources don't get created properly and I get errors like:
`Couldn't deploy version 1: ReplicationController "gogs-1" is invalid: spec.template.metadata.labels: Invalid value: {"app":"jenkins","deployment":"gogs-1","deploymentconfig":"gogs"}: `selector` does not match template `labels``

However if I use `oc process -f /path/template.yaml | oc create -f -` then it gets instantiated correctly and works fine.

##### Version
oc v3.2.0.20
kubernetes v1.2.0-36-g4a3f9c5

**Comment by bparees** [comment]
@csrwng ptal.

**Comment by csrwng** [comment]
I get similar errors with new-app and process:
```
# oc process -f /tmp/cicd-gogs-template.yaml
error: bufio.Scanner: token too long

# oc new-app -f /tmp/cicd-gogs-template.yaml
error: unable to load template file "/tmp/cicd-gogs-template.yaml": bufio.Scanner: token too long
error: no match for "/tmp/cicd-gogs-template.yaml"

The 'oc new-app' command will match arguments to the following types:

  1. Images tagged into image streams in the current project or the 'openshift' project
     - if you don't specify a tag, we'll add ':latest'
  2. Images in the Docker Hub, on remote registries, or on the local Docker engine
  3. Templates in the current project or the 'openshift' project
  4. Git repository URLs or local paths that point to Git repositories

--allow-missing-images can be used to point to an image that does not exist yet.

See 'oc new-app -h' for examples.
```

**Comment by csrwng** [comment]
```
# oc version
oc v1.3.0-alpha.1-259-g8cd1645
kubernetes v1.3.0-alpha.1-331-g0522e63
```

**Comment by csrwng** [comment]
Actually scratch that, wrong file

**Comment by csrwng** [comment]
Testing with the right file, both 'oc new-app' and 'oc process' work for me using the latest released origin client (v1.3.0-alpha.1). Will look into the OSE 3.2 version.

**Comment by csrwng** [comment]
I'm not able to reproduce with the latest v3.2:
```
# oc version
oc v3.2.0.46-1-g3fab54d
kubernetes v1.2.0-36-g4a3f9c5
```

**Comment by bparees** [comment]
What does work mean in this context? I got the impression from the issue
report that the problem was not actually instantiating the template, it was
that the deployment that resulted from the DC failed in one case and not
the other. Did your stuff fully deploy successfully in both cases?

Ben Parees | OpenShift

**Comment by csrwng** [comment]
ahh, then I was looking at the wrong thing... let me try again

**Comment by csrwng** [comment]
ok so the reason it's behaving differently with new-app is that new-app will set labels on all produced objects, and not all in the same way. 

When labels are not specified, new-app decides that it will label all resources with 'app=[name]' where [name] is the first name it finds in the objects/images it needs to deploy. The template itself already contains several app=[blah] labels throughout. 

In some fields, new-app will overwrite the labels included in template, while in others, it will not. For example, it will not overwrite object metadata labels:
https://github.com/openshift/origin/blob/master/pkg/util/labels.go#L47

but it will overwrite deployment config nested labels:
https://github.com/openshift/origin/blob/master/pkg/util/labels.go#L174

The resulting deployment config is not consistent. I know this issue will be fixed in @mfojtik's rebase because now we will always be overwriting labels. In the meantime, you can get around the issue by specifying labels on new-app so that it doesn't use the default 'app=[blah]':

```
oc new-app /path/template.yaml -l application=gogs
```

**Comment by siamaksade** [comment]
I see. Will this behaviour (overwriting labels) in `oc new-app` be consistent with `oc process -f file.yaml | oc create -f -`?

**Comment by jorgemoralespou** [comment]
@csrwng Let me understand future behavior:

If I have a template that has a label of "application=example".

* If I do `oc new-app template.yaml` what will I get as application label in the resulting objects?
* If I do `oc new-app template.yaml --name=myapp` what will I get as application label in the resulting objects?
* If I do `oc new-app template.yaml -l application=otherapp` what will I get as application label in the resulting objects?
* If I do `oc new-app template.yaml --name=myapp -l application=otherapp` what will I get as application label in the resulting objects?   

Thanks for answering.

**Comment by csrwng** [comment]
> I see. Will this behaviour (overwriting labels) in oc new-app be consistent with oc process -f file.yaml | oc create -f -?

Not quite. It's a "feature" of new-app to set a common label to all the objects it creates, while `oc create -f` is strictly passthrough. I'm not saying this is the way that is should be, but that's what it is right now.

**Comment by csrwng** [comment]
@jorgemoralespou 

> If I do oc new-app template.yaml what will I get as application label in the resulting objects?

The same application=example, plus app=[some name] that new-app comes up with. Note that only the `app` label is special for new-app. Which is why @siamaksade was seeing an issue. His template already includes `app` labels.


> If I do oc new-app template.yaml --name=myapp what will I get as application label in the resulting objects?

You'll get existing labels + name=myapp

> If I do oc new-app template.yaml -l application=otherapp what will I get as application label in the resulting objects?

Your existing labels (application=example) will be overwritten with `application=otherapp` throughout all objects that exist in your template.



> If I do oc new-app template.yaml --name=myapp -l application=otherapp what will I get as application label in the resulting objects?

Same as in the previous question... your existing label will be overwritten by `application=otherapp`

**Comment by jorgemoralespou** [comment]
@csrwng 

So, I think I should reformulate my question, if the template as a label of `app` and you create an application through new-app, app label will be replaced? Is this documented? 

Is `app`the label we'll be using to standardise grouping in overview?

**Comment by csrwng** [comment]
@jorgemoralespou 

Yes, the existing app label on your template will be replaced (iff you don't specify your own labels for new-app with `-l mylabel=blah`).

This is currently not documented, and as I said before, the behavior in the code has been inconsistent up to now. When the rebase lands we'll need to update the documentation to explain the behavior.

> Is appthe label we'll be using to standardise grouping in overview?

I am not aware that it has a special meaning outside of new-app yet, but @jwforres or @spadgett can maybe comment on that.

**Comment by spadgett** [comment]
> I am not aware that it has a special meaning outside of new-app yet, but @jwforres or @spadgett can maybe comment on that.

We show the app label if it's set in the new overview. Service grouping isn't controlled by the app label, though. It is set through annotations on the services. For example,

https://github.com/openshift/origin/blob/master/examples/jenkins/pipeline/pipelinetemplate.json#L62-L64

This is temporary until we have true service linking.

**Comment by siamaksade** [comment]
The new UI uses the service object's _app_ label as the group name near the route url which makes _new-app_ replacing the _app_ label even more troublesome. We are basically telling the user _new-app_ is for creating apps from templates, but don't use it if you have multiple groups in the same template. 

Using _oc new-app_:
<img width="974" alt="new-app" src="https://cloud.githubusercontent.com/assets/1174934/18203317/94b77fec-7116-11e6-8363-016a12d0968f.png">

Using _oc process_:
<img width="948" alt="process" src="https://cloud.githubusercontent.com/assets/1174934/18203320/9887a35e-7116-11e6-894d-38868f1b5485.png">

_new-app_ should assign the _app_ label if it doesn't exist and it shouldn't override it if already defined by the template.

**Comment by jwforres** [comment]
I agree, if the template already has an app label defined on the template labels, we should not be overwriting it.  @bparees @csrwng

**Comment by jwforres** [comment]
But in this case you were putting the app labels on the individual services, will need to think about this.

**Comment by csrwng** [comment]
@jwforres this was a change that came through in a kube rebase, @mfojtik ?

**Comment by liggitt** [comment]
thread at https://github.com/openshift/origin/pull/8856#discussion-diff-66240013

discussion during that review settled on the new-app label trumping what was in the template. we can revisit, but quickly

/cc @smarterclayton

**Comment by liggitt** [comment]
and see test ensuring the new (disputed) behavior at https://github.com/openshift/origin/blob/master/test/cmd/newapp.sh#L65-L69

**Comment by smarterclayton** [comment]
There is a difference in new-app between the explicit label (user sets --labels) and implicit label (we add app to something based on the name).

If the template already has the label app=SOMETHINGELSE it seems wrong to change it on implicit, but ok to change it on explicit.

**Comment by csrwng** [comment]
@smarterclayton - how would that work? If we find a conflict in any of the objects in the template, we don't apply the label to any of them?
Or do we only apply it to objects that don't already have an app label?

**Comment by smarterclayton** [comment]
Probably whatever would cause the least disruption.  What would cause the
least disruption?

**Comment by csrwng** [comment]
My thinking is that we don't set it on any object if you find the label on at least one... otherwise you risk making them inconsistent ... some objects with one app label value and others with a different value.

**Comment by smarterclayton** [comment]
So for templates we are creating we should never set app.  I guess I'm ok
with setting none (on implicit) if none are set.  Explicit label set still
needs to overwrite.  We also need to allow someone to bypass the implicit
label set (probably via an empty label selector).

**Comment by bparees** [comment]
new-app currently treats all objects it creates (Whether generated by it, or derived from a template) equally.  Handling template objects distinctly from other objects is going to be annoying.

The current behavior is:

1) Look at the labels specified by the user on the command line.
2) if "app" isn't one of those labels, create a new "app" label as if the user specified it (w/ a computed value).
3) Take all the objects produced by new-app (including anything rendered by a template, in which case the object will have any labels the template itself specified as part of the Labels block, and anything the object definition within the template had) and apply all the labels from (1) and (2) to those objects, overwriting any existing label keys.

Obviously we want overwrite behavior for user defined labels, probably in all cases.
There's an argument that we don't want overwrite behavior for the "app" label, ever.

But i strongly dislike the idea that if someone supplies a template with 3 objects that have an "app" label and 2 that don't, they're going to end up with 5 objects that have 2 different "app" label values.  Always overwriting ensures that doesn't happen.

**Comment by csrwng** [comment]
@bparees we would not apply the implicit label if it conflicts with ANY resource being created.

**Comment by bparees** [comment]
@csrwng meaning we'd scan *all* the objects, check if *any* of them have an "app" label, and if so, don't apply the app label to anything?

**Comment by csrwng** [comment]
@bparees basically 1) copy the original set of objects, 2) merge the implicit label with the error-on-conflict flag set (don't remember what it was called) 3) If a conflict is found, keep the original list

**Comment by bparees** [comment]
@csrwng tomayto tomahto

**Comment by jwforres** [comment]
Whatever you guys decide, we have to do it in the console as well.

**Comment by bparees** [comment]
I don't see us doing this for 3.3 unless someone is screaming incredibly loudly.

**Comment by jwforres** [comment]
Agreed

**Comment by jorgemoralespou** [comment]
I don't agree with setting app automatically if the user is not adviced of the fact, in the implicit scenario. 
An example is that I'm creating an app, myapp, I deploy 2 templates for that, Wildfly and mongodb. In implicit mode I'll have 2 applications. 

If app is something we are making somehow first class citizen, we need to always have an explicit behavior where user knows what's happening, as I don't think that is ready for a user to cascade changes to all created objects to adjust the labels in the wrongly created objects to match having both deployments (and everything related) belonging to the same app.
